From 5bbe2294121f904a5dcb697bb751445bf837c352 Mon Sep 17 00:00:00 2001
From: Robby Cai <robby.cai@nxp.com>
Date: Fri, 25 Dec 2015 14:04:54 +0800
Subject: [PATCH] MLK-12060 media: output: pxp_v4l2: Add filter function for
 dma_request_channel

Without the filter function, the dma_request_channel() API could possibly get
a sdma channel. The dma_request_channel() API does not return error but the
consequent operations on that channel will be invalid then the upper level
application expects to be stuck.

The patch added the filter functions to get right dma channel for the fix.

Signed-off-by: Robby Cai <robby.cai@nxp.com>
(cherry picked from commit be335d5154e0b66a68f3dad0501f9f3637c8f0d4)
---
 drivers/media/platform/mxc/output/mxc_pxp_v4l2.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/drivers/media/platform/mxc/output/mxc_pxp_v4l2.c b/drivers/media/platform/mxc/output/mxc_pxp_v4l2.c
index 22840da..0f0d86b 100644
--- a/drivers/media/platform/mxc/output/mxc_pxp_v4l2.c
+++ b/drivers/media/platform/mxc/output/mxc_pxp_v4l2.c
@@ -37,6 +37,7 @@
 #include <linux/delay.h>
 #include <linux/console.h>
 #include <linux/mxcfb.h>
+#include <linux/platform_data/dma-imx.h>
 
 #include <media/videobuf-dma-contig.h>
 #include <media/v4l2-common.h>
@@ -249,6 +250,14 @@ static void video_dma_done(void *arg)
 	spin_unlock(&pxp->lock);
 }
 
+static bool chan_filter(struct dma_chan *chan, void *arg)
+{
+	if (imx_dma_is_pxp(chan))
+		return true;
+	else
+		return false;
+}
+
 static int acquire_dma_channel(struct pxps *pxp)
 {
 	dma_cap_mask_t mask;
@@ -270,7 +279,7 @@ static int acquire_dma_channel(struct pxps *pxp)
 	dma_cap_zero(mask);
 	dma_cap_set(DMA_SLAVE, mask);
 	dma_cap_set(DMA_PRIVATE, mask);
-	chan = dma_request_channel(mask, NULL, NULL);
+	chan = dma_request_channel(mask, chan_filter, NULL);
 	if (!chan)
 		return -EBUSY;
 
-- 
1.8.0

